src/Utilities/PooledMemory.h

```diff
@@ -0,0 +1,244 @@
+//////////////////////////////////////////////////////////////////////////////////////
+// This file is distributed under the University of Illinois/NCSA Open Source License.
+// See LICENSE file in top directory for details.
+//
+// Copyright (c) 2016 Jeongnim Kim and QMCPACK developers.
+//
+// File developed by: Jeongnim Kim, jeongnim.kim@intel.com, Intel Corp.
+//                    Ye Luo, yeluo@anl.gov, Argonne National Laboratory
+//
+// File created by: Ye Luo, yeluo@anl.gov, Argonne National Laboratory
+//////////////////////////////////////////////////////////////////////////////////////
+
+
+#ifndef QMCPLUSPLUS_POOLEDMEMORY_H
+#define QMCPLUSPLUS_POOLEDMEMORY_H
+
+#include <vector>
+#include <complex>
+#include <limits>
+#include "simd/allocator.hpp"
+
+namespace qmcplusplus
+{
+
+/** Memory pool to manage arrays+scalars with alignment
+ * @tparam T_scalar : type of the scalar data type, typically OHMMS_PRECISION_FULL
+ * @tparam PageSize : page size in bytes, default=4096 (4K)
+ * @tparam Alloc : allocator, Mallocator<T, PageSize>
+ *
+ * The bulk part is accessed directly by address
```

Comment:
I assuming these design points:
 - Reuse sections of the buffer directly as storage in some wavefunction classes (saves memory and copy overhead).  The attach methods were added to some container classes to enable this.
 - Because the buffer storage is used directly, maintaining alignment is important
 - Data should be stored in a single contiguous container, making it easy to pass the data as a single MPI message (otherwise it would be simpler to use separate structures for block and scalar data)

Because the new reuse of buffer data, I am concerned about issues due to object lifetimes - that the wavefunction classes could access buffer locations after the buffer has been deallocated or reused.
What is the lifetime?  It appears copyFromBuffer attaches the storage from the buffer.  But when is the buffer data no longer needed?  When copyFromBuffer is called again on a new walker.
It might be possible to implement some sort of reference counting scheme to detect this case.

I'm also concerned about the variability in how the attach and resetByRef methods handle freeing existing memory.  resetByRef will free existing memory, but Vector::attach will throw an exception if previously allocated memory is present (meaning the user needs to free the memory before calling attach)

Reply:
1) Yes.
2) Yes.
3) PooledData was the old implementation mixing block and scalar data. I decide to keep supporting block and scalar for backward compatibility.

In any QMC driver, copyFromBuffer is always called before any pbyp moves on wavefunction classes. So the buffer location is always correct. updateBuffer can be called as many as you want.

The buffer is persistent with the walker. Once created, always valid until the walker is destroyed.
In DMC, load balancing is the last operation in an iteration, the next cycle will redo the copyFromBuffer ensuring correct assignment.

resetByRef was introduced earlier than attach. I think we should change its name as well as behaviour to throwing an exception.